### PR TITLE
Clean up Azure Hosting APIs

### LINKED ISSUE #6174: Clean up Azure Hosting APIs
We should do an API Review for the Azure Hosting APIs. Specifically:

- Remove the Experimental callbacks, for example
https://github.com/dotnet/aspire/blob/a8b1a619d67549243342dfd2c92cfaf26905cb65/src/Aspire.Hosting.Azure.Storage/AzureStorageExtensions.cs#L42-L43
    - Instead, callers will call `builder.AddAzureStorage("storage").ConfigureConstruct(c => ...)` if they want to customize the Azure resource

- Rename `ResourceModuleConstruct` to `AzureResourceInfrastructure` since it inherits from `Azure.Provisioning.Infrastructure`.

- Rename `AzureConstructResource` to `AzureProvisioningResource`

- Rename `ConfigureConstruct` to `ConfigureInfrastructure`
    - This matches the Azure.Provisioning naming of this type.

We should also take a holistic view of all our Azure Hosting APIs and do an API Review on them to make sure all the things we are shipping for 9.0 make sense for the long term.

cc @mitchdenny @davidfowl @tg-msft

### PR DESCRIPTION
**Pull request by eerhardt**
## Description

See https://spa.apiview.dev/review/2c5cd25605b44759b267536b0fcecc84?activeApiRevisionId=3caa6b0d2c314f3f8b388e2b86443057&nId=Aspire.Hosting.Azure.AzureBicepResourceAnnotation for the API review of the current assembly.

Contributes to #6174

## Checklist

- Is this feature complete?
  - [x] No. A couple more API changes are expected for 9.0.0
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
- Did you add public API?
  - [x] Yes
    - If yes, did you have an API Review for it?
      - [x] Yes
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [x] Yes
- Does the change make any security assumptions or guarantees?
  - [x] No
- Does the change require an update in our Aspire docs?
  - [x] Yes
    - Link to aspire-docs issue: https://github.com/dotnet/docs-aspire/issues/1652

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6304)

### CODE REVIEW

**Comment by eerhardt** [comment]
The only thing from the API review that isn't done yet is to remove the following properties:

```C#
public class AzureResourceInfrastructure : Infrastructure { 
   public ProvisioningParameter PrincipalIdParameter { get; }
   public ProvisioningParameter PrincipalNameParameter { get; }
   public ProvisioningParameter PrincipalTypeParameter { get; }
```

I will do that in a follow up PR, since the change is pretty involved and this PR is big enough by itself.

File: src/Aspire.Hosting.Azure/AzureResourceInfrastructure.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+using Azure.Provisioning.Expressions;
+using Azure.Provisioning;
+
+namespace Aspire.Hosting.Azure;
+
+/// <summary>
+/// An Azure Provisioning <see cref="Infrastructure" /> which represents the root Bicep module that is generated for an Azure resource.
+/// </summary>
+public class AzureResourceInfrastructure : Infrastructure
```
Comment (eerhardt) [inline src/Aspire.Hosting.Azure/AzureResourceInfrastructure.cs:12]:
```suggestion
public sealed class AzureResourceInfrastructure : Infrastructure
```

File: src/Aspire.Hosting.Azure/AzureResourceInfrastructure.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            Description = "The location for the resource(s) to be deployed.",
+            Value = BicepFunction.GetResourceGroup().Location
+        });
+    }
+
+    /// <summary>
+    /// The Aspire <see cref="AzureProvisioningResource"/> resource that this <see cref="AzureResourceInfrastructure"/> represents.
+    /// </summary>
+    public AzureProvisioningResource Resource { get; }
```
Comment (eerhardt) [inline src/Aspire.Hosting.Azure/AzureResourceInfrastructure.cs:31]: Should this be named `AspireResource`?

* ModelResource
* DistributedResource
* AzureProvisioningResource

?

Comment (davidfowl) [reply src/Aspire.Hosting.Azure/AzureResourceInfrastructure.cs:31]: AspireResource

**Review by davidfowl — APPROVED** [review]

**Pull request merged**